drivers/video/hm0360.c

```diff
@@ -0,0 +1,885 @@
+	{HM0360_CCI_ROI_START_END_H, 0xF0},
+	{HM0360_CCI_ROI_START_END_V, 0xD0},
+	{HM0360_CCI_COMMAND_UPDATE, 0x01},
+};
+
+#define HM0360_VIDEO_FORMAT_CAP(width, height, format)                                             \
+	{                                                                                          \
+		.pixelformat = (format), .width_min = (width), .width_max = (width),               \
+		.height_min = (height), .height_max = (height), .width_step = 0, .height_step = 0  \
+	}
+
+static const struct video_format_cap fmts[] = {
+	HM0360_VIDEO_FORMAT_CAP(160, 120, VIDEO_PIX_FMT_GREY), /* QQVGA  */
+	HM0360_VIDEO_FORMAT_CAP(320, 240, VIDEO_PIX_FMT_GREY), /* QVGA  */
+	HM0360_VIDEO_FORMAT_CAP(640, 480, VIDEO_PIX_FMT_GREY), /* VGA   */
+	{0}};
+
+static int hm0360_get_caps(const struct device *dev, struct video_caps *caps)
+{
+	caps->format_caps = fmts;
+
+	return 0;
+}
+
+static int hm0360_set_fmt(const struct device *dev, struct video_format *fmt)
+{
+	const struct hm0360_config *config = dev->config;
+	struct hm0360_data *data = dev->data;
+	int ret;
+	size_t idx;
+
+	/* Note: hm0360 only supports grayscale */
+	LOG_DBG("HM01B0 set_fmt: %d x %d, fmt: %s", fmt->width, fmt->height,
+		VIDEO_FOURCC_TO_STR(fmt->pixelformat));
+
+	ret = video_format_caps_index(fmts, fmt, &idx);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (!memcmp(&data->fmt, fmt, sizeof(data->fmt))) {
+		/* nothing to do */
+		return 0;
+	}
+
+	memcpy(&data->fmt, fmt, sizeof(data->fmt));
+
+	switch (fmts[idx].width_min) {
+	case 160: /* QQVGA */
+		ret = video_write_cci_multiregs(&config->bus, hm0360_qqvga_regs,
+						ARRAY_SIZE(hm0360_qqvga_regs));
+		break;
+	case 320: /* QVGA */
+		ret = video_write_cci_multiregs(&config->bus, hm0360_qvga_regs,
+						ARRAY_SIZE(hm0360_qvga_regs));
+		break;
+	case 640: /* VGA */
+		ret = video_write_cci_multiregs(&config->bus, hm0360_vga_regs,
+						ARRAY_SIZE(hm0360_vga_regs));
+		break;
+	default:
+		CODE_UNREACHABLE;
+	}
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static int hm0360_get_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct hm0360_data *data = dev->data;
+
+	memcpy(fmt, &data->fmt, sizeof(data->fmt));
+
+	return 0;
+}
+
+static int hm0360_set_frmival(const struct device *dev, struct video_frmival *frmival)
+{
+	const struct hm0360_config *config = dev->config;
+	struct video_format fmt;
+	struct hm0360_data *drv_data = dev->data;
+	int ret;
+
+	uint32_t pll_cfg = 0;
+	uint32_t osc_div = 0;
+	bool highres = false;
+
+	ret = hm0360_get_fmt(dev, &fmt);
+	if (ret < 0) {
+		LOG_ERR("Can not get format!");
+		return ret;
+	}
+
+	if (fmt.width == 640 && fmt.height == 480) {
+		highres = true;
+	}
+
+	if (frmival->numerator <= 10) {
+		osc_div = highres ? 0x02 : 0x03;
+	} else if (frmival->numerator <= 30) {
+		osc_div = highres ? 0x01 : 0x02;
+	} else {
+		/* Set to the max possible FPS at this resolution. */
+		osc_div = highres ? 0x00 : 0x01;
+	}
```

Comment:
If someone asks for a frame interval of 11 FPS, best is to set 10 FPS rather than 30.

But I think you are already aware of where to find some code for it. :]

- https://github.com/zephyrproject-rtos/zephyr/pull/98154#discussion_r2463811631

Thanks for the look around

Reply:
Fixed to match how https://github.com/zephyrproject-rtos/zephyr/pull/98154#discussion_r2463811631 does it.